request.go

```diff
@@ -27,3 +56,40 @@ func (r *request) encode(pe packetEncoder) (err error) {
 	}
 	return pe.pop()
 }
+
+func decodeRequest(pd packetDecoder) (reqHeader *requestHeader, req interface{}, err error) {
+	reqHeader = &requestHeader{}
+	if err = reqHeader.decode(pd); err != nil {
+		return nil, nil, err
+	}
+	decoder := makeDecoder(reqHeader.key, reqHeader.version)
+	if decoder == nil {
+		return reqHeader, nil, unknownKey(fmt.Errorf("Unknown request key: %d", reqHeader.key))
+	}
+	if err = decoder.decode(pd); err != nil {
+		return reqHeader, nil, err
+	}
+	return reqHeader, decoder, nil
+}
+
+func makeDecoder(key, version int16) decoder {
+	switch key {
+	case 0:
+		return &ProduceRequest{}
+	case 1:
+		return &FetchRequest{}
+	case 2:
+		return &OffsetRequest{}
+	case 3:
+		return &MetadataRequest{}
+	case 8:
+		return &OffsetCommitRequest{Version: version}
+	case 9:
+		return &OffsetFetchRequest{}
+	case 10:
+		return &ConsumerMetadataRequest{}
+	}
+	return nil
+}
+
+type unknownKey error
```

Comment:
No it cannot. None of the changes that I made are used in production. This particular error is not needed. We can just do with `fmt.Errorf`. In fact it is never return even in tests. Because all requests implement decoder interface now. If a new request is introduced by a developer who forgets to provide `decode` function then this error will be returned, but only in tests.